revolutionpi/digital_interrupts.go

```diff
@@ -0,0 +1,187 @@
+)
+
+type digitalInterrupt struct {
+	boardWorkers *sync.WaitGroup
+	interrupt    pinwrappers.ReconfigurableDigitalInterrupt
+	line         *gpio.LineWithEvent
+	cancelCtx    context.Context
+	cancelFunc   func()
+	config       *board.DigitalInterruptConfig
+
+	Name         string // Variable name
+	Address      uint16 // Address of the byte in the process image
+	Length       uint16 // length of the variable in bits. Possible values are 1, 8, 16 and 32
+	ControlChip  *gpioChip
+	outputOffset uint16
+	inputOffset  uint16
+}
+
+// func (b *Board) createDigitalInterrupt(
+// 	ctx context.Context,
+// 	config board.DigitalInterruptConfig,
+// 	gpioMappings map[string]GPIOBoardMapping,
+// 	// If we are reconfiguring a board, we might already have channels subscribed and listening for
+// 	// updates from an old interrupt that we're creating on a new pin. In that case, reuse the part
+// 	// that holds the callbacks.
+// 	oldCallbackHolder pinwrappers.ReconfigurableDigitalInterrupt,
+// ) (*digitalInterrupt, error) {
+// 	mapping, ok := gpioMappings[config.Pin]
+// 	if !ok {
+// 		return nil, errors.Errorf("unknown interrupt pin %s", config.Pin)
+// 	}
+
+// 	chip, err := gpio.OpenChip(mapping.GPIOChipDev)
+// 	if err != nil {
+// 		return nil, err
+// 	}
+// 	defer utils.UncheckedErrorFunc(chip.Close)
+
+// 	line, err := chip.OpenLineWithEvents(
+// 		uint32(mapping.GPIO), gpio.Input, gpio.BothEdges, "viam-interrupt")
+// 	if err != nil {
+// 		return nil, err
+// 	}
+
+// 	var interrupt pinwrappers.ReconfigurableDigitalInterrupt
+// 	if oldCallbackHolder == nil {
+// 		interrupt, err = pinwrappers.CreateDigitalInterrupt(config)
+// 		if err != nil {
+// 			return nil, multierr.Combine(err, line.Close())
+// 		}
+// 	} else {
+// 		interrupt = oldCallbackHolder
+// 		if err := interrupt.Reconfigure(config); err != nil {
+// 			return nil, err // Should never have errors, but this makes the linter happy
+// 		}
+// 	}
+
+//		cancelCtx, cancelFunc := context.WithCancel(ctx)
+//		result := digitalInterrupt{
+//			boardWorkers: &b.activeBackgroundWorkers,
+//			interrupt:    interrupt,
+//			line:         line,
+//			cancelCtx:    cancelCtx,
+//			cancelFunc:   cancelFunc,
+//			config:       &config,
+//		}
+//		result.startMonitor()
+//		return &result, nil
+//	}
+func (di *digitalInterrupt) initialize() error {
+	di.ControlChip.logger.Info("yo dev path: ", di.ControlChip.dev)
+	chip, err := gpio.OpenChip("piControl0")
+	if err != nil {
+		return err
+	}
+	defer utils.UncheckedErrorFunc(chip.Close)
+
+	di.ControlChip.logger.Info("yo chip: ", chip)
+	line, err := chip.OpenLineWithEvents(
+		uint32(di.Address), gpio.Input, gpio.BothEdges, "viam-interrupt")
+	if err != nil {
+		return err
+	}
+
+	cfg := board.DigitalInterruptConfig{Name: di.Name, Pin: di.Name}
+	interrupt, err := pinwrappers.CreateDigitalInterrupt(cfg)
+	if err != nil {
+		return multierr.Combine(err, line.Close())
+	}
+	di.line = line
+	di.interrupt = interrupt
+	di.config = &cfg
```

Comment:
attempted to use the same interrupt setup in generic linux to create the event handler, but the `gpio` package (understandably) assumes that each gpio pin has their own address/offset. The gpio values on the rev pi are all stored within a 2 byte word, and  the `gpio` package does not allow me to pass in the correct bit. 

I also do not have a way to pull out the event handler functions from the `gpio` package, as they are all labeled as internal and trying to reimplement the entire package seemed excessive. 

For now I am going to leave out the Streamticks/tick functionality for interrupts. Basically just going to implement Value and call it a day

Reply:
Agree, thank you for looking into this - if we need it, then we'll implement using rev pi's polling functionality.